### PR TITLE
v3

### LINKED ISSUE #44: iter.Seq and naming changes
It would be nice to have the following methods available:

```go
func (m *OrderedMap[K, V]) All() iter.Seq2[K, V] {...}
func (m *OrderedMap[K, V]) Keys() iter.Seq[K] {...}
func (m *OrderedMap[K, V]) Values() iter.Seq[V] {...}
```

These would line up nicely with the newly implemented [`All`](https://pkg.go.dev/maps#All), [`Keys`](https://pkg.go.dev/maps#Keys) and [`Values`](https://pkg.go.dev/maps#Values) functions in the standard `maps` package. Adding the two `iter.Seq` methods also allows for compatibility with [`slices.Collect`](https://pkg.go.dev/slices#Collect).

Unfortunately, `Keys()` conflicts with the existing pre-iterators method and would require a major version change (or a different name). For the long-term I think the package would benefit from this consistency with the standard library. Also, the current version of the `Keys()` method can be inefficient on large maps and moving to an iterator by default instead is a good API improvement IMO.

Some other points:

- `All()` would be identical to the current `Iterator()` method, but again, would be more idiomatic.
- `Values()` was never implemented previously so no problems adding this.
- As we currently have `ReverseIterator()` maybe we should have `AllFromFront()` and `AllFromBack()` instead. This is both idiomatic and consistent with the naming in this package. `All()` could be an alias for `AllFromFront()`.

I understand that a major API change isn't always desirable, but I believe these changes will benefit the design in the long-term. I've created a branch with the proposed changes [in my fork](https://github.com/pd93/orderedmap/commit/d80adfe0db1369a3b5bc3c8148c4362385833229).

### PR DESCRIPTION
**Pull request by pd93**
Closes #44

As discussed in #44, this PR proposes a couple of breaking changes (and therefore a major version bump to v3):

- **BREAKING** - `m.Keys()` now returns an `iter.Seq` iterator instead of a list of keys.
  - This should be much more performant on large maps.
  - It is also more consistent with `maps.Keys` from the standard library.
- Added `m.Values()` which returns an `iter.Seq` iterator.
  - Also consistent with `maps.Values` from the standard library.
- **BREAKING** - Renamed `Iterator()` method to `AllFromFront()`.
- **BREAKING** - Renamed `ReverseIterator()` to `AllFromBack()`.
- Added `All()` method aliased to `AllFromFront()`.
  - This is for consistency with the standard library's `slices.All` and `maps.All`.
- Updated minimum version to 1.23
  - This also meant we can move the iterator methods into the main `orderedmap.go` file as they no longer need to be constrained by a build tag
- Updated tests
- Updated README
- Updated CI to run on v3 and Go 1.23

I have submitted this PR in 2 commits. The first just directly copied the `v2` code into the new `v3` directory. The second commit contains the v3 changes. This should more easily allow you to see and review the actual changes.

- - -
This change is [Reviewable](https://reviewable.io/reviews/elliotchance/orderedmap/46)

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -9,13 +9,11 @@ jobs:
...
+          - "^1.23"
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
       - uses: actions/setup-go@v3
         with:
           go-version: ${{ matrix.go }}
-      - run: go test -v ./...
-      - run: cd v2 && go test -v ./...
+      - run: cd v3 && go test -v ./...
```
Comment (elliotchance) [inline .github/workflows/ci.yaml:19]: This shouldn't replace v2, we should run CI on both.

File: README.md [hunk 2]
```diff
@@ -1,12 +1,12 @@
-# 🔃 github.com/elliotchance/orderedmap/v2 [![GoDoc](https://godoc.org/github.com/elliotchance/orderedmap/v2?status.svg)](https://godoc.org/github.com/elliotchance/orderedmap/v2)
+# 🔃 github.com/elliotchance/orderedmap/v3 [![GoDoc](https://godoc.org/github.com/elliotchance/orderedmap/v2?status.svg)](https://godoc.org/github.com/elliotchance/orderedmap/v2)
```
Comment (elliotchance) [inline README.md:1]: Update the doc URLs as well.

File: v3/orderedmap.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+// Len returns the number of elements in the map.
+func (m *OrderedMap[K, V]) Len() int {
+	return len(m.kv)
+}
+
+// All is an alias for AllFromFront. It returns an iterator that yields all
+// elements in the map starting at the front (oldest Set element).
+func (m *OrderedMap[K, V]) All() iter.Seq2[K, V] {
```
Comment (elliotchance) [inline v3/orderedmap.go:95]: We can remove this, there's no need for two methods that do the same thing.

**Review by elliotchance — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (pd93) [reply .github/workflows/ci.yaml:21]: Wasn't sure on the best strategy here. We can't run v3 against 1.18, but I assume we still want to run v1/2 against it. I think we would need separate tests to do this. Let me know if you want this or if you're happy with just running all 3 versions against 1.23.

File: v3/orderedmap.go (hunk 3 above)
Comment (pd93) [reply v3/orderedmap.go]: The thinking here was to align with `maps.All` and `slices.All`, but happy with just the `AllFromFront` and `AllFromBack` methods.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (elliotchance) [reply .github/workflows/ci.yaml:21]: You can just duplicate the job itself. Rename this job to `unit-tests-v2` and add a `unit-tests-v3`.

**Review by elliotchance — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -19,26 +19,56 @@ func main() {
...
 
 ## Iterating
 
-Be careful using `Keys()` as it will create a copy of all of the keys so it's
-only suitable for a small number of items:
+The following methods all return
+[iterators](https://go.dev/doc/go1.23#iterators) that can be used to loop over
+elements in an ordered map:
+
+- `All()` - (Aliased to `AllFromFront`)
```
Comment (elliotchance) [inline README.md:37]: Remove this line and fix the reference to the snippet below.

File: v3/orderedmap.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	ll list[K, V]
+}
+
+func NewOrderedMap[K comparable, V any]() *OrderedMap[K, V] {
+	return &OrderedMap[K, V]{
+		kv: make(map[K]*Element[K, V]),
+	}
+}
+
+func NewOrderedMapWithElements[K comparable, V any](els ...*Element[K, V]) *OrderedMap[K, V] {
```
Comment (elliotchance) [inline v3/orderedmap.go:16]: Can you also include `NewOrderedMapWithCapacity` from this recent PR: #48

**Review by elliotchance — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (pd93) [reply README.md]: Whoops, fixed

File: v3/orderedmap.go (hunk 5 above)
Comment (pd93) [reply v3/orderedmap.go:24]: Rebased and added the changes from #48 into the first commit

**Review by elliotchance — APPROVED** [review]

**Comment by elliotchance** [comment]
Thank you @pd93 for this awesome work! It's now released as v3.0.0.

**Pull request merged**